# Basic Features

In [12]:
import numpy as np
import pandas as pd
%matplotlib inline

In [13]:
from pandas import read_csv

In [14]:
series1 = read_csv('../Data/daily-minimum-temperatures.csv', 
                     header=0, index_col=0, parse_dates=True,squeeze=True)

In [15]:
series2 = pd.read_csv('../Data/daily-minimum-temperatures.csv', 
                     header=0, index_col=0, parse_dates=True,squeeze=True)

In [16]:
type(series1) # This is a Pandas Series

pandas.core.series.Series

In [17]:
type(series2) # this is a Pandas Dataframe

pandas.core.series.Series

In [23]:
series1.head()

Date
1981-01-01    20.7
1981-01-02    17.9
1981-01-03    18.8
1981-01-04    14.6
1981-01-05    15.8
Name: Temp, dtype: float64

In [24]:
series2.head()

Date
1981-01-01    20.7
1981-01-02    17.9
1981-01-03    18.8
1981-01-04    14.6
1981-01-05    15.8
Name: Temp, dtype: float64

Creating a data Frame

In [22]:
from pandas import DataFrame 
dataframe=DataFrame()

In [25]:
dataframe['month']=[series1.index[i].month for i in range(len(series1))]

In [27]:
dataframe['day']=[series1.index[i].day for i in range(len(series1))]

In [28]:
dataframe['temperature']=[series1[i] for i in range(len(series1))]

In [32]:
dataframe.head(5)

month  day  temperature
0      1    1         20.7
1      1    2         17.9
2      1    3         18.8
3      1    4         14.6
4      1    5         15.8

In [33]:
from pandas import concat 


In [37]:
DataFrame(series1.values)

0
0     20.7
1     17.9
2     18.8
3     14.6
4     15.8
...    ...
3645  14.0
3646  13.6
3647  13.5
3648  15.7
3649  13.0

[3650 rows x 1 columns]

In [38]:
temp=DataFrame(series1.values)

In [52]:
dataframe2=concat([temp.shift(1),temp],axis=1)
dataframe2.columns = ['t', 't+1']

temp.shift(1) generate a new variable that is shifted the dataset temp by 1 create the column and adding NaN value for the first row. Therefore, i f i define this new variable a t column, then the original can be interpreted as t+1


In [53]:
dataframe2.head()

t   t+1
0   NaN  20.7
1  20.7  17.9
2  17.9  18.8
3  18.8  14.6
4  14.6  15.8

In [46]:
temp2=DataFrame(series1.values)

In [48]:
dataframe3=concat([temp2.shift(3),temp2.shift(2), temp2.shift(1), temp2], axis=1)
dataframe3.columns=['t-2','t-1','t','t+1']

In [51]:
dataframe3.head()

t-2   t-1     t   t+1
0   NaN   NaN   NaN  20.7
1   NaN   NaN  20.7  17.9
2   NaN  20.7  17.9  18.8
3  20.7  17.9  18.8  14.6
4  17.9  18.8  14.6  15.8

Some Statistics:

In [57]:
temp.head()

0
0  20.7
1  17.9
2  18.8
3  14.6
4  15.8

In [72]:
shifted=temp.shift(1)
shifted.head()

0
0   NaN
1  20.7
2  17.9
3  18.8
4  14.6

In [73]:
window=shifted.rolling(window=2)

In [80]:
means=window.mean()
means.head()

0
0    NaN
1    NaN
2  19.30
3  18.35
4  16.70

In [83]:
dataframe4=concat([means, temp],axis=1)
dataframe4.columns=['mean(t-1,t)','t+1']

In [84]:
dataframe4.head()

mean(t-1,t)   t+1
0          NaN  20.7
1          NaN  17.9
2        19.30  18.8
3        18.35  14.6
4        16.70  15.8

Note the first NaN comes from the shift of the series, the second Nan is because the cannot be used to calculate a mean value. 19.30 is the mean between 20.7 and 17.9

In [87]:
width=3
shifted=temp.shift(width-1)
window=shifted.rolling(window=width)
df=concat([window.min(), window.mean(), window.max(),temp],axis=1)
df.columns=['min','mean','max','t+1']

In [88]:
df.head()

min       mean   max   t+1
0   NaN        NaN   NaN  20.7
1   NaN        NaN   NaN  17.9
2   NaN        NaN   NaN  18.8
3   NaN        NaN   NaN  14.6
4  17.9  19.133333  20.7  15.8

In [89]:
(20.7+17.9+18.8)/3

19.13333333333333

In [92]:
window=temp.expanding()
window

Expanding [min_periods=1,center=False,axis=0]

In [93]:
df2=concat([window.min(), window.mean(), window.max(),temp.shift(-1)],axis=1)
df2.columns=['min','mean','max','t+1']

In [94]:
df2.head()

min       mean   max   t+1
0  20.7  20.700000  20.7  17.9
1  17.9  19.300000  20.7  18.8
2  17.9  19.133333  20.7  14.6
3  14.6  18.000000  20.7  15.8
4  14.6  17.560000  20.7  15.8